activerecord/lib/active_record/relation/batches.rb

```diff
@@ -45,14 +46,21 @@ module Batches
     #
     # NOTE: You can't set the limit either, that's used to control
     # the batch sizes.
-    def find_each(start: nil, batch_size: 1000)
+    def find_each(start: nil, end_at: nil, batch_size: 1000)
       if block_given?
-        find_in_batches(start: start, batch_size: batch_size) do |records|
+        find_in_batches(start: start, end_at: end_at, batch_size: batch_size) do |records|
           records.each { |record| yield record }
         end
       else
-        enum_for(:find_each, start: start, batch_size: batch_size) do
-          start ? where(table[primary_key].gteq(start)).size : size
+        enum_for(:find_each, start: start, batch_size: batch_size, end_at: end_at) do
+          relation = self
+          if start || end_at
+            relation = relation.where(table[primary_key].gteq(start)) if start
```

Comment:
I think we can extract this code to a private method. It is duplicated in three places